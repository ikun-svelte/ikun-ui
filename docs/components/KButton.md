---
title: KButton
lang: en-US
---

# KButton

Commonly used button.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/button
```

```bash [yarn]
yarn add @ikun-ui/button
```

```bash [npm]
npm install @ikun-ui/button
```

:::

## Basic usage

Use `type`, `round` and `circle` to define Button's style.

<demo src="button/basic.svelte"  github='Button'></demo>

## Plain Button

When you set the `plain` property to `true`, plain buttons will be rendered.

<demo src="button/plain.svelte"  github='Button'></demo>

## Ghost Button

When you set the `ghost` property to `true`, ghost buttons will be rendered.

::: tip
Ghost Button takes precedence over Plain Button
:::

<demo src="button/ghost.svelte"  github='Button'></demo>

## Disabled Button

Use `disabled` attribute to determine whether a button is disabled.  
It accepts a `Boolean` value.

<demo src="button/disabled.svelte" github='Button'></demo>

## Icon Button

Use the icon attribute to add icon.

<demo src="button/icon.svelte" github='Button'></demo>

## Button Size

Use the `size` property to control the button size.
It supports enumerations such as `md`, `sm`, and `lg`.

<demo src="button/size.svelte" github='Button'></demo>

## Button Border

Use the `isBorder` property to determine if the button is bordered.

::: tip
Applies only to plain button
:::

<demo src="button/isBorder.svelte" github='Button'></demo>

## Button Group

Displayed as a button group, can be used to group a series of similar operations.

Use tag `<KButtonGroup>` to group your buttons.

<demo src="button/group.svelte" github='ButtonGroup'></demo>

## Link Button

Use the `to` attribute to set a specific url that you can click to jump to.

<demo src="button/link.svelte" github='Button'></demo>

## Button Props

| Name     | Type                                                 | Default    | Description                                               |
| -------- | ---------------------------------------------------- | ---------- | --------------------------------------------------------- |
| type     | `primary` / `success` / `error` / `warning` / `info` | `primary'` | Button type                                               |
| size     | `'sm' / 'md' / 'lg'`                                 | `md`       | Button size                                               |
| icon     | `string`                                             | `-`        | The class name of the icon, following the unocss standard |
| iconSize | `number`                                             | `-`        | The icon size of the button                               |
| to       | `string`                                             | `-`        | Determine whether it's a link button                      |
| round    | `string` / `number`                                  | `-`        | Button's border radius                                    |
| circle   | `boolean`                                            | `false`    | Determine whether it's a circle button                    |
| isBorder | `boolean`                                            | `false`    | Border style button                                       |
| plain    | `boolean`                                            | `false`    | Determine whether it's a plain button                     |
| ghost    | `boolean`                                            | `false`    | Determine whether it's a ghost button                     |
| disabled | `boolean`                                            | `false`    | Disable the button                                        |
| cls      | `string`                                             | `-`        | Additional class for component                            |
| attrs    | `Record<string, string>`                             | `{}`       | Additional attributes                                     |

## Button Events

| Name  | Description                            | Type                |
| ----- | -------------------------------------- | ------------------- |
| click | Event fired when the button is clicked | `(e: Event)=> void` |

## Button Slots

| Name    | Description               |
| ------- | ------------------------- |
| default | Customize default content |

## Button Group Props

| Name     | Type                                                 | Default | Description                                                         |
| -------- | ---------------------------------------------------- | ------- | ------------------------------------------------------------------- |
| type     | `primary` / `success` / `error` / `warning` / `info` | `false` | Buttons type.                                                       |
| size     | `'sm' / 'md' / 'lg'`                                 | `md`    | Buttons size. It will cover button size.                            |
| iconSize | `number`                                             | `-`     | The icon size of the buttons. It will cover button iconSize.        |
| isBorder | `boolean`                                            | `false` | Border style buttons. If true, will override the button's isBorder. |
| disabled | `boolean`                                            | `false` | Disable buttons. If true, will override the button's disabled.      |
| cls      | `string`                                             | `-`     | Additional class for component                                      |
| attrs    | `Record<string, string>`                             | `{}`    | Additional attributes                                               |
