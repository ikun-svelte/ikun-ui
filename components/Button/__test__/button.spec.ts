import { tick } from 'svelte';
import { afterEach, expect, test, vi, describe, beforeEach } from 'vitest';
import KButton from '../src';
import KButtonDisabled from './fixtrue/disabled.svelte';

let host;

const initHost = () => {
	host = globalThis.document.createElement('div');
	host.setAttribute('id', 'host');
	globalThis.document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KButton', () => {
	test('props: to', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				to: 'https://github.com/ikun-svelte/ikun-ui'
			}
		});
		expect(instance).toBeTruthy();
		expect(host.querySelector('a')?.getAttribute('href')).toBe(
			'https://github.com/ikun-svelte/ikun-ui'
		);
		expect(host.innerHTML.endsWith('</a>')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: round', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				round: 30
			}
		});
		expect(instance).toBeTruthy();
		expect(host.firstChild!.style.borderRadius).toBe('30px');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: icon', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				icon: 'icon-carbon-settings'
			}
		});
		expect(instance).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-icon--base k-icon-transition icon-carbon-settings k-button--primary__icon k-button--primary__icon__fill'
			)
		).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: circle', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				circle: true
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--circle')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: type', async () => {
		//@ts-ignore
		const success = new KButton({
			target: host,
			props: {
				type: 'success'
			}
		});
		expect(success).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--success__fill k-button--success__hover__fill k-button--success__active k-button--success__focus'
			)
		).toBeTruthy();
		//@ts-ignore
		const error = new KButton({
			target: host,
			props: {
				type: 'error'
			}
		});
		expect(error).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--error__fill k-button--error__hover__fill k-button--error__active k-button--error__focus'
			)
		).toBeTruthy();
		//@ts-ignore
		const warning = new KButton({
			target: host,
			props: {
				type: 'warning'
			}
		});
		expect(warning).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--warning__fill k-button--warning__hover__fill k-button--warning__active k-button--warning__focus'
			)
		).toBeTruthy();
		//@ts-ignore
		const info = new KButton({
			target: host,
			props: {
				type: 'info'
			}
		});
		expect(info).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--info__fill k-button--info__hover__fill k-button--info__active k-button--info__focus'
			)
		).toBeTruthy();
		//@ts-ignore
		const primary = new KButton({
			target: host,
			props: {
				type: 'primary'
			}
		});
		expect(primary).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus'
			)
		).toBeTruthy();

		expect(host.innerHTML).matchSnapshot();
	});

	test('props: plain', async () => {
		//@ts-ignore
		const success = new KButton({
			target: host,
			props: {
				type: 'success',
				plain: true
			}
		});
		expect(success).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--success k-button--success__hover k-button--success__active k-button--success__focus'
			)
		).toBeTruthy();
		//@ts-ignore
		const error = new KButton({
			target: host,
			props: {
				type: 'error',
				plain: true
			}
		});
		expect(error).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--error k-button--error__hover k-button--error__active k-button--error__focus'
			)
		).toBeTruthy();
		//@ts-ignore
		const warning = new KButton({
			target: host,
			props: {
				type: 'warning',
				plain: true
			}
		});
		expect(warning).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--warning k-button--warning__hover k-button--warning__active k-button--warning__focus'
			)
		).toBeTruthy();
		//@ts-ignore
		const info = new KButton({
			target: host,
			props: {
				type: 'info',
				plain: true
			}
		});
		expect(info).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--info k-button--info__hover k-button--info__active k-button--info__focus'
			)
		).toBeTruthy();
		//@ts-ignore
		const primary = new KButton({
			target: host,
			props: {
				type: 'primary',
				plain: true
			}
		});
		expect(primary).toBeTruthy();
		expect(
			host!.innerHTML.includes(
				'k-button--primary k-button--primary__hover k-button--primary__active k-button--primary__focus'
			)
		).toBeTruthy();

		expect(host.innerHTML).matchSnapshot();
	});

	test('props: ghost', async () => {
		//@ts-ignore
		const success = new KButton({
			target: host,
			props: {
				type: 'success',
				ghost: true
			}
		});
		expect(success).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--success__ghost')).toBeTruthy();

		//@ts-ignore
		const error = new KButton({
			target: host,
			props: {
				type: 'error',
				ghost: true
			}
		});
		expect(error).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--error__ghost')).toBeTruthy();
		//@ts-ignore
		const warning = new KButton({
			target: host,
			props: {
				type: 'warning',
				ghost: true
			}
		});
		expect(warning).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--warning__ghost')).toBeTruthy();
		//@ts-ignore
		const info = new KButton({
			target: host,
			props: {
				type: 'info',
				ghost: true
			}
		});
		expect(info).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--info__ghost')).toBeTruthy();
		//@ts-ignore
		const primary = new KButton({
			target: host,
			props: {
				type: 'primary',
				ghost: true
			}
		});
		expect(primary).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--primary__ghost')).toBeTruthy();

		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				disabled: true
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-cur-disabled k-button--disabled')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: disabled value change', async () => {
		//@ts-ignore
		const instance = new KButtonDisabled({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-cur-disabled k-button--disabled')).not.toBeTruthy();
		const btn = host.querySelector('.k-button');
		btn.click();
		await tick();
		expect(host!.innerHTML.includes('k-cur-disabled k-button--disabled')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: button md size', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				icon: 'i-carbon-settings',
				size: 'md'
			}
		});
		expect(instance).toBeTruthy();
		expect(!host!.innerHTML.includes('k-button--sm')).toBeTruthy();
		expect(!host!.innerHTML.includes('k-button--lg')).toBeTruthy();
		expect(host!.innerHTML.includes('i-carbon-settings')).toBeTruthy();
		expect(host!.innerHTML.includes('style="width: 16px; height: 16px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: button sm size', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				icon: 'i-carbon-settings',
				size: 'sm'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--sm')).toBeTruthy();
		expect(host!.innerHTML.includes('i-carbon-settings')).toBeTruthy();
		expect(host!.innerHTML.includes('style="width: 12px; height: 12px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: button lg size', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				icon: 'i-carbon-settings',
				size: 'lg'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--lg')).toBeTruthy();
		expect(host!.innerHTML.includes('i-carbon-settings')).toBeTruthy();
		expect(host!.innerHTML.includes('style="width: 20px; height: 20px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: isBorder', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				isBorder: true,
				type: 'primary'
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('k-button--primary__border')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: iconSize', async () => {
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				icon: 'i-carbon-settings',
				iconSize: 10086
			}
		});
		expect(instance).toBeTruthy();
		expect(host!.innerHTML.includes('i-carbon-settings')).toBeTruthy();
		expect(host!.innerHTML.includes('style="width: 10086px; height: 10086px;"')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('event: should trigger click event', async () => {
		const mockFn = vi.fn();
		//@ts-ignore
		const instance = new KButton({
			target: host
		});
		await tick();
		//@ts-ignore
		instance.$on('click', mockFn);
		const btn = host.getElementsByTagName('button')[0];
		btn.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(instance).toBeTruthy();
		expect(mockFn).toBeCalled();
	});

	test('event: should not trigger click event when disabled', async () => {
		const mockFn = vi.fn();
		//@ts-ignore
		const instance = new KButton({
			target: host,
			props: {
				disabled: true
			}
		});
		await tick();
		//@ts-ignore
		instance.$on('click', mockFn);
		const btn = host.getElementsByTagName('button')[0];
		btn.click(); // or btn.dispatchEvent(new window.Event('click', { bubbles: true }))
		await tick();
		expect(instance).toBeTruthy();
		expect(mockFn).not.toBeCalled();
	});
});
