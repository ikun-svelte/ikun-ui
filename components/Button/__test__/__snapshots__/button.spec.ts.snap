// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KButton > props: button lg size 1`] = `"<button style=\\"border-radius: 4px; font-size: 20px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--lg\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-settings k-button--primary__icon k-button--primary__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 20px; height: 20px;\\"></span>   </button>"`;

exports[`Test: KButton > props: button md size 1`] = `"<button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-settings k-button--primary__icon k-button--primary__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span>   </button>"`;

exports[`Test: KButton > props: button sm size 1`] = `"<button style=\\"border-radius: 4px; font-size: 12px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--sm\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-settings k-button--primary__icon k-button--primary__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 12px; height: 12px;\\"></span>   </button>"`;

exports[`Test: KButton > props: circle 1`] = `"<button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--circle\\" aria-hidden=\\"true\\">   </button>"`;

exports[`Test: KButton > props: disabled 1`] = `"<button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-cur-disabled k-button--disabled\\" aria-hidden=\\"true\\">   </button>"`;

exports[`Test: KButton > props: ghost 1`] = `"<button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--success__ghost k-button--success__active k-button--success__focus k-button--success__border\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--error__ghost k-button--error__active k-button--error__focus k-button--error__border\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--warning__ghost k-button--warning__active k-button--warning__focus k-button--warning__border\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--info__ghost k-button--info__active k-button--info__focus k-button--info__border\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__ghost k-button--primary__active k-button--primary__focus k-button--primary__border\\" aria-hidden=\\"true\\">   </button>"`;

exports[`Test: KButton > props: icon 1`] = `"<button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition icon-carbon-settings k-button--primary__icon k-button--primary__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 16px; height: 16px;\\"></span>   </button>"`;

exports[`Test: KButton > props: iconSize 1`] = `"<button style=\\"border-radius: 4px; font-size: 10086px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus\\" aria-hidden=\\"true\\"><span class=\\"k-icon--base k-icon-transition i-carbon-settings k-button--primary__icon k-button--primary__icon__fill\\" role=\\"\\" aria-hidden=\\"true\\" style=\\"width: 10086px; height: 10086px;\\"></span>   </button>"`;

exports[`Test: KButton > props: isBorder 1`] = `"<button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus k-button--primary__border\\" aria-hidden=\\"true\\">   </button>"`;

exports[`Test: KButton > props: plain 1`] = `"<button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--success k-button--success__hover k-button--success__active k-button--success__focus\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--error k-button--error__hover k-button--error__active k-button--error__focus\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--warning k-button--warning__hover k-button--warning__active k-button--warning__focus\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--info k-button--info__hover k-button--info__active k-button--info__focus\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary k-button--primary__hover k-button--primary__active k-button--primary__focus\\" aria-hidden=\\"true\\">   </button>"`;

exports[`Test: KButton > props: round 1`] = `"<button style=\\"border-radius: 30px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus\\" aria-hidden=\\"true\\">   </button>"`;

exports[`Test: KButton > props: to 1`] = `"<a style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus\\" aria-hidden=\\"true\\" href=\\"https://github.com/ikun-svelte/ikun-ui\\">   </a>"`;

exports[`Test: KButton > props: type 1`] = `"<button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--success__fill k-button--success__hover__fill k-button--success__active k-button--success__focus\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--error__fill k-button--error__hover__fill k-button--error__active k-button--error__focus\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--warning__fill k-button--warning__hover__fill k-button--warning__active k-button--warning__focus\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--info__fill k-button--info__hover__fill k-button--info__active k-button--info__focus\\" aria-hidden=\\"true\\">   </button><button style=\\"border-radius: 4px; font-size: 16px;\\" class=\\"k-button k-button--base k-button--primary__fill k-button--primary__hover__fill k-button--primary__active k-button--primary__focus\\" aria-hidden=\\"true\\">   </button>"`;
